README.md
# tap-shopify

This is a [Singer](https://singer.io) tap that produces JSON-formatted data
following the [Singer
spec](https://github.com/singer-io/getting-started/blob/master/SPEC.md).

This tap:

- Pulls raw data from [Shopify](https://help.shopify.com/en/api/reference)
- Extracts the following resources:
  - [Abandoned Checkouts](https://help.shopify.com/en/api/reference/orders/abandoned_checkouts)
  - [Collects](https://help.shopify.com/en/api/reference/products/collect)
  - [Custom Collections](https://help.shopify.com/en/api/reference/products/customcollection)
  - [Customers](https://help.shopify.com/en/api/reference/customers)
  - [Metafields](https://help.shopify.com/en/api/reference/metafield)
  - [Orders](https://help.shopify.com/en/api/reference/orders)
  - [Products](https://help.shopify.com/en/api/reference/products)
  - [Transactions](https://help.shopify.com/en/api/reference/orders/transaction)
  - [Locations](https://help.shopify.com/en/api/reference/inventory/location)
  - [Inventory Levels](https://help.shopify.com/en/api/reference/inventory/inventorylevel)
  - [Inventory Item](https://help.shopify.com/en/api/reference/inventory/inventoryitem)
- Outputs the schema for each resource
- Incrementally pulls data based on the input state
- When Metafields are selected, this tap will sync the Shopify store's top-level Metafields and any additional Metafields for selected tables that also have them (ie: Orders, Products, Customers)

## Quick Start

1. Install

   poetry install

2. Create the config file

   Create a JSON file called `config.json`. Its contents should look like:

   ```json
   {
     "start_date": "2010-01-01",
     "api_key": "<Shopify API Key>",
     "shop": "test_shop",
     "request_timeout": 300
   }
   ```

   The `start_date` specifies the date at which the tap will begin pulling data
   (for those resources that support this).

   The `api_key` is the API key for your Shopify shop generated via an OAuth flow.

   The `shop` is your Shopify shop which will be the value `test_shop` in the string `https://test_shop.myshopify.com`

   The `request_timeout` is the timeout for the requests. Default: 300 seconds

3. Run the Tap in Discovery Mode

   tap-shopify -c config.json -d

   See the Singer docs on discovery mode
   [here](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#discovery-mode).

4. Run the Tap in Sync Mode

   poetry run tap-shopify --config tap.config.json --properties tap.catalog.json

## Troubleshooting

If you see the error

```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (\_ssl.c:1131)
```

Then for macOS search for `Install Certificates.command` and run it

---

Copyright &copy; 2019 Stitch


CHANGELOG.md
# Changelog

## ## 1.6.3

- Dependabot bump pyjwt to 2.4.0

## ## 1.6.2

- Merge latest from singer-io/tap-shopify

## ## 1.1.10

- Extracting from just `metafields` or `abandoned_checkouts` tables will no longer cause an error.

## 1.6.1

- Fixing Tranformation Issues [#149] (https://github.com/singer-io/tap-shopify/pull/149)

## 1.6.0

- API/SDK Upgrade to v10.0.0 [#135] (https://github.com/singer-io/tap-shopify/pull/135)
- New Field Additions to Schema [#140] (https://github.com/singer-io/tap-shopify/pull/140)

## 1.5.1

- Request Timeout Implementation [#129](https://github.com/singer-io/tap-shopify/pull/129)

## 1.5.0

- Adds `events` stream [#127](https://github.com/singer-io/tap-shopify/pull/127)

## 1.4.0

- Add shop info in record [#115](https://github.com/singer-io/tap-shopify/pull/115)
- Add inventory item data [#118] (https://github.com/singer-io/tap-shopify/pull/118)
- Add Locations stream and TDL-13614: Add Inventory Levels stream [#114] (https://github.com/singer-io/tap-shopify/pull/114)
- Added best practices [#116] (https://github.com/singer-io/tap-shopify/pull/116)
- Discover mode should check token [#120] (https://github.com/singer-io/tap-shopify/pull/120)

## 1.3.6

- Fixes a bug in the canonicalize function for a 'receipt' key existing with a null value [#119](https://github.com/singer-io/tap-shopify/pull/119)

## 1.3.5

- Add `status` field to `Products` stream [#108](https://github.com/singer-io/tap-shopify/pull/108)

## 1.3.4

- Correct abandoned_checkouts schema to correctly reflect some properties as arrays [#44](https://github.com/singer-io/tap-shopify/pull/44)

## 1.3.3.

- Add `build` to the list of fields we canonicalize for the Transactions stream [#103](https://github.com/singer-io/tap-shopify/pull/103)

## 1.3.2

- Add `python_requires` to `setup.py` [#101](https://github.com/singer-io/tap-shopify/pull/101)
  - We've tested the tap on `python 3.5.2` and `python 3.8.0`

## 1.3.1

- Canonicalize `Timestamp` to `timestamp` on `Transactions.receipt` [#98](https://github.com/singer-io/tap-shopify/pull/98)

## 1.3.0

This version ships both [#96][pr#96] and [#97][pr#97].

From [#97][pr#97]:

- removes the "untestable streams" list from all tests.
- makes the test match the tap and expect the default page size to be 175, not 250
- adds bookmarking to order_refunds
- adds bookmarking to transactions
- adds shopify error handling to transactions
  - The tests would fail with unhandled 429s
- adds pagination to transactions

From [#96][pr#96]:

- Update the API version from `2020-10` to `2021-04`

[pr#97]: https://github.com/singer-io/tap-shopify/pull/97/
[pr#96]: https://github.com/singer-io/tap-shopify/pull/96

## 1.2.10

- Add `null, object` to customer schema definition [#94](https://github.com/singer-io/tap-shopify/pull/94)

## 1.2.9

- Bumps `singer-python` from `5.11.0` to `5.12.1` [#91](https://github.com/singer-io/tap-shopify/pull/91)

## 1.2.8

- Modified schema so that all fields using `multipleOf` are now using `singer.decimal` [#88](https://github.com/singer-io/tap-shopify/pull/88)

## 1.2.7

- Change how exceptions are logged to make the error messages more consistent [#84](https://github.com/singer-io/tap-shopify/pull/84)

## 1.2.6

- Accepts any string for `accepts_marketing_updated_at` field on the `customers` stream [#69](https://github.com/singer-io/tap-shopify/pull/69)

## 1.2.5

- Bumps `singer-python` from `5.4.1` to `5.9.1` [#67](https://github.com/singer-io/tap-shopify/pull/67)

## 1.2.4

- Adds `accepts_marketing_updated_at` to shared `customer` schema [#61](https://github.com/singer-io/tap-shopify/pull/61/)

## 1.2.3

- Bumped Shopify API version to 2020-07 (SDK version 8.0.1) [#63](https://github.com/singer-io/tap-shopify/pull/63/)

## 1.2.2

- Fixes issue where `products` returns 0 records due to a change in the `status` parameter, Shopify now requires `published_status` [#59](https://github.com/singer-io/tap-shopify/pull/59)

## 1.2.1

- Update the line_item schema to allow an object under the properties key [#58](https://github.com/singer-io/tap-shopify/pull/58)

## 1.2.0

- Bump ShopifyAPI version from 3.1.0 -> 7.0.1 [#54](https://github.com/singer-io/tap-shopify/pull/54/)
- Explicitly specify Shopify API version as `2019-10` [#54](https://github.com/singer-io/tap-shopify/pull/54/)

## 1.1.17

- Use try/except around JSON metafield's data and fallback to string [#50](https://github.com/singer-io/tap-shopify/pull/50)

## 1.1.16

- Handles null/non-integer values for `results_per_page` [#48](https://github.com/singer-io/tap-shopify/pull/48)

## 1.1.15

- Lowering the page size from 250 to 175 as per recommendation from
  Shopify to mitigate receiving 500s [#45](https://github.com/singer-io/tap-shopify/pull/45)

## 1.1.14

- Bump minimum value for numbers to 1e-10 [commit](https://github.com/singer-io/tap-shopify/commit/a2abf49be96b07f80610d63c514241f829780dcf)

## 1.1.12

- Canonicalize `transaction__receipt__version` to always prefer and coerce to `version` [#41](https://github.com/singer-io/tap-shopify/pull/41)

## 1.1.11

- Add `http_request_timer` metrics to HTTP requests [#39](https://github.com/singer-io/tap-shopify/pull/39)

## 1.1.10

- Canonicalize `transaction__receipt__token` to always use `token` instead of `Token` [#37](https://github.com/singer-io/tap-shopify/pull/37)

## 1.1.9

- Retry pattern will now fall back to lowercase if `Retry-After` not present [#35](https://github.com/singer-io/tap-shopify/pull/35)

## 1.1.8

- Uses patternProperties to match extra fields on transactions receipts [#33](https://github.com/singer-io/tap-shopify/pull/33)

## 1.1.7

- Add `results_per_page` as a config param and allow float values for `date_window_size` [#30](https://github.com/singer-io/tap-shopify/pull/30)

## 1.1.6

- Fix bookmark resetting for `since_id` after date window finishes [#29](https://github.com/singer-io/tap-shopify/pull/29)

## 1.1.5


tap_shopify/__init__.py
#!/usr/bin/env python3
import json
import os
import copy

import pyactiveresource
import shopify
import singer
from singer import Transformer, metadata, utils
from tap_shopify.context import Context
from tap_shopify.exceptions import ShopifyError
from tap_shopify.streams.base import shopify_error_handling, get_request_timeout

REQUIRED_CONFIG_KEYS = ["shop", "api_key"]
LOGGER = singer.get_logger()
SDC_KEYS = {'id': 'integer', 'name': 'string', 'myshopify_domain': 'string'}

@shopify_error_handling
def initialize_shopify_client():
    api_key = Context.config['api_key']
    shop = Context.config['shop']
    version = '2022-01'
    LOGGER.info("APIKEY: " + str(api_key))
    session = shopify.Session(shop, version, api_key)
    shopify.ShopifyResource.activate_session(session)

    # set request timeout
    shopify.Shop.set_timeout(get_request_timeout())

    # Shop.current() makes a call for shop details with provided shop and api_key
    return shopify.Shop.current().attributes

def get_abs_path(path):
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

# Load schemas from schemas folder
def load_schemas():
    schemas = {}

    # This schema represents many of the currency values as JSON schema
    # 'number's, which may result in lost precision.
    for filename in os.listdir(get_abs_path('schemas')):
        path = get_abs_path('schemas') + '/' + filename
        schema_name = filename.replace('.json', '')
        with open(path) as file:
            schemas[schema_name] = json.load(file)

    return schemas


def get_discovery_metadata(stream, schema):
    mdata = metadata.new()
    mdata = metadata.write(mdata, (), 'table-key-properties', stream.key_properties)
    mdata = metadata.write(mdata, (), 'forced-replication-method', stream.replication_method)

    if stream.replication_key:
        mdata = metadata.write(mdata, (), 'valid-replication-keys', [stream.replication_key])

    for field_name in schema['properties'].keys():
        if field_name in stream.key_properties or field_name == stream.replication_key:
            mdata = metadata.write(mdata, ('properties', field_name), 'inclusion', 'automatic')
        else:
            mdata = metadata.write(mdata, ('properties', field_name), 'inclusion', 'available')

    return metadata.to_list(mdata)

def load_schema_references():
    shared_schema_file = "definitions.json"
    shared_schema_path = get_abs_path('schemas/')

    refs = {}
    with open(os.path.join(shared_schema_path, shared_schema_file)) as data_file:
        refs[shared_schema_file] = json.load(data_file)

    return refs

def add_synthetic_key_to_schema(schema):
    for k in SDC_KEYS:
        schema['properties']['_sdc_shop_' + k] = {'type': ["null", SDC_KEYS[k]]}
    return schema

def discover():
    initialize_shopify_client() # Checking token in discover mode

    raw_schemas = load_schemas()
    streams = []

    refs = load_schema_references()
    for schema_name, schema in raw_schemas.items():
        if schema_name not in Context.stream_objects:
            continue

        stream = Context.stream_objects[schema_name]()

        # resolve_schema_references() is changing value of passed refs.
        # Customer is a stream and it's a nested field of orders and abandoned_checkouts streams
        # and those 3 _sdc fields are also added inside nested field customer for above 2 stream
        # so create a copy of refs before passing it to resolve_schema_references().
        refs_copy = copy.deepcopy(refs)
        catalog_schema = add_synthetic_key_to_schema(
            singer.resolve_schema_references(schema, refs_copy))

        # create and add catalog entry
        catalog_entry = {
            'stream': schema_name,
            'tap_stream_id': schema_name,
            'schema': catalog_schema,
            'metadata': get_discovery_metadata(stream, schema),
            'key_properties': stream.key_properties,
            'replication_key': stream.replication_key,
            'replication_method': stream.replication_method,
            'column_order': [str(column) for column in catalog_schema['properties']]
        }
        streams.append(catalog_entry)

    return {'streams': streams}

def shuffle_streams(stream_name):
    '''
    Takes the name of the first stream to sync and reshuffles the order
    of the list to put it at the top
    '''
    matching_index = 0
    for i, catalog_entry in enumerate(Context.catalog["streams"]):
        if catalog_entry["tap_stream_id"] == stream_name:
            matching_index = i
    top_half = Context.catalog["streams"][matching_index:]
    bottom_half = Context.catalog["streams"][:matching_index]
    Context.catalog["streams"] = top_half + bottom_half

# pylint: disable=too-many-locals
def sync():
    shop_attributes = initialize_shopify_client()
    sdc_fields = {"_sdc_shop_" + x: shop_attributes[x] for x in SDC_KEYS}

    # Emit all schemas first so we have them for child streams
    for stream in Context.catalog["streams"]:
        if Context.is_selected(stream["tap_stream_id"]):
            singer.write_schema(stream["tap_stream_id"],
                                stream["schema"],
                                stream["key_properties"],
                                bookmark_properties=stream["replication_key"])
            Context.counts[stream["tap_stream_id"]] = 0

    # If there is a currently syncing stream bookmark, shuffle the
    # stream order so it gets sync'd first
    currently_sync_stream_name = Context.state.get('bookmarks', {}).get('currently_sync_stream')
    if currently_sync_stream_name:
        shuffle_streams(currently_sync_stream_name)

    # Loop over streams in catalog
    for catalog_entry in Context.catalog['streams']:
        stream_id = catalog_entry['tap_stream_id']
        stream = Context.stream_objects[stream_id]()

        if not Context.is_selected(stream_id):
            LOGGER.info('Skipping stream: %s', stream_id)
            continue

        LOGGER.info('Syncing stream: %s', stream_id)

        if not Context.state.get('bookmarks'):
            Context.state['bookmarks'] = {}
        Context.state['bookmarks']['currently_sync_stream'] = stream_id

        # some fields have epoch-time as date, hence transform into UTC date
        with Transformer(singer.UNIX_SECONDS_INTEGER_DATETIME_PARSING) as transformer:
            for rec in stream.sync():
                extraction_time = singer.utils.now()
                record_schema = catalog_entry['schema']
                record_metadata = metadata.to_map(catalog_entry['metadata'])
                rec = transformer.transform({**rec, **sdc_fields},
                                            record_schema,
                                            record_metadata)
                singer.write_record(stream_id,
                                    rec,
                                    time_extracted=extraction_time)
                Context.counts[stream_id] += 1

        Context.state['bookmarks'].pop('currently_sync_stream')
        singer.write_state(Context.state)

    LOGGER.info('----------------------')
    for stream_id, stream_count in Context.counts.items():
        LOGGER.info('%s: %d', stream_id, stream_count)
    LOGGER.info('----------------------')

@utils.handle_top_exception(LOGGER)
def main():
    try:
        # Parse command line arguments
        args = utils.parse_args(REQUIRED_CONFIG_KEYS)

        Context.config = args.config
        Context.state = args.state

        # If discover flag was passed, run discovery mode and dump output to stdout
        if args.discover:
            catalog = discover()
            print(json.dumps(catalog, indent=2))
        # Otherwise run in sync mode
        else:
            Context.tap_start = utils.now()
            if args.catalog:
                # Sort the properties
                streams = args.properties['streams']
                for stream in streams:
                    new_properties = {}
                    old_properties = stream['schema']['properties']
                    order = stream['column_order']

                    for column in order:
                        new_properties[column] = old_properties[column]

                    stream['schema']['properties'] = new_properties

                Context.catalog = args.properties
            else:
                Context.catalog = discover()

            sync()
    except pyactiveresource.connection.ResourceNotFound as exc:
        raise ShopifyError(exc, 'Ensure shop is entered correctly') from exc
    except pyactiveresource.connection.UnauthorizedAccess as exc:
        raise ShopifyError(exc, 'Invalid access token - Re-authorize the connection') \
            from exc
    except pyactiveresource.connection.ConnectionError as exc:
        msg = ''
        try:
            body_json = exc.response.body.decode()
            body = json.loads(body_json)
            msg = body.get('errors')
        finally:
            raise ShopifyError(exc, msg) from exc
    except Exception as exc:
        raise ShopifyError(exc) from exc

if __name__ == "__main__":
    main()


tap_shopify/streams/order_refunds.py
import shopify
from singer.utils import strftime, strptime_to_utc
from tap_shopify.context import Context
from tap_shopify.streams.base import (Stream,
                                      shopify_error_handling,
                                      OutOfOrderIdsError)

class OrderRefunds(Stream):
    name = 'order_refunds'
    replication_object = shopify.Refund
    replication_key = 'created_at'

    @shopify_error_handling()
    def get_refunds(self, parent_object, since_id):
        # set timeout
        self.replication_object.set_timeout(self.request_timeout)
        return self.replication_object.find(
            order_id=parent_object.id,
            limit=self.results_per_page,
            since_id=since_id,
            order='id asc')

    def get_objects(self):
        selected_parent = Context.stream_objects['orders']()
        selected_parent.name = "refund_orders"

        # Page through all `orders`, bookmarking at `refund_orders`
        for parent_object in selected_parent.get_objects():
            since_id = 1
            while True:
                refunds = self.get_refunds(parent_object, since_id)
                for refund in refunds:
                    if refund.id < since_id:
                        raise OutOfOrderIdsError("refund.id < since_id: {} < {}".format(
                            refund.id, since_id))
                    yield refund
                if len(refunds) < self.results_per_page:
                    break
                if refunds[-1].id != max([o.id for o in refunds]):
                    raise OutOfOrderIdsError("{} is not the max id in refunds ({})".format(
                        refunds[-1].id, max([o.id for o in refunds])))
                since_id = refunds[-1].id

    def sync(self):
        bookmark = self.get_bookmark()
        max_bookmark = bookmark
        for refund in self.get_objects():
            refund_dict = refund.to_dict()
            replication_value = strptime_to_utc(refund_dict[self.replication_key])
            if replication_value >= bookmark:
                yield refund_dict

            if replication_value > max_bookmark:
                max_bookmark = replication_value

        self.update_bookmark(strftime(max_bookmark))


Context.stream_objects['order_refunds'] = OrderRefunds


tap_shopify/streams/base.py
import datetime
import functools
import math
import sys
import socket
import backoff
import pyactiveresource
import pyactiveresource.formats
import simplejson
import singer
from singer import metrics, utils
from tap_shopify.context import Context
import time

LOGGER = singer.get_logger()

RESULTS_PER_PAGE = 175

# set default timeout of 300 seconds
REQUEST_TIMEOUT = 300

# We've observed 500 errors returned if this is too large (30 days was too
# large for a customer)
DATE_WINDOW_SIZE = 1

# We will retry a 500 error a maximum of 5 times before giving up
MAX_RETRIES = 5

# function to return request timeout
def get_request_timeout():

    request_timeout = REQUEST_TIMEOUT # set default timeout
    timeout_from_config = Context.config.get('request_timeout')
    # updated the timeout value if timeout is passed in config and not from 0, "0", ""
    if timeout_from_config and float(timeout_from_config):
        # update the request timeout for the requests
        request_timeout = float(timeout_from_config)

    return request_timeout

def is_not_status_code_fn(status_code):
    def gen_fn(exc):
        if getattr(exc, 'code', None) and exc.code not in status_code:
            return True
        # Retry other errors up to the max
        return False
    return gen_fn

def leaky_bucket_handler(details):
    LOGGER.info("Received 429 -- sleeping for %s seconds",
                details['wait'])

def retry_handler(details):
    LOGGER.info("Received 500 or retryable error -- Retry %s/%s",
                details['tries'], MAX_RETRIES)

#pylint: disable=unused-argument
def retry_after_wait_gen(**kwargs):
    # This is called in an except block so we can retrieve the exception
    # and check it.
    exc_info = sys.exc_info()
    resp = exc_info[1].response
    # Retry-After is an undocumented header. But honoring
    # it was proven to work in our spikes.
    # It's been observed to come through as lowercase, so fallback if not present
    sleep_time_str = resp.headers.get('Retry-After', resp.headers.get('retry-after'))
    yield math.floor(float(sleep_time_str))

# boolean function to check if the error is 'timeout' error or not
def is_timeout_error(error_raised):
    """
        This function checks whether the error contains 'timed out' substring and return boolean
        values accordingly, to decide whether to backoff or not.
    """
    # retry if the error string contains 'timed out'
    if str(error_raised).__contains__('timed out'):
        return False
    return True

def shopify_error_handling(fnc):
    @backoff.on_exception(backoff.expo, # timeout error raise by Shopify
                          (pyactiveresource.connection.Error, socket.timeout),
                          giveup=is_timeout_error,
                          max_tries=MAX_RETRIES,
                          factor=2)
    @backoff.on_exception(backoff.expo,
                          (pyactiveresource.connection.ServerError,
                           pyactiveresource.formats.Error,
                           simplejson.scanner.JSONDecodeError),
                          giveup=is_not_status_code_fn(range(500, 599)),
                          on_backoff=retry_handler,
                          max_tries=MAX_RETRIES)
    @backoff.on_exception(retry_after_wait_gen,
                          pyactiveresource.connection.ClientError,
                          giveup=is_not_status_code_fn([429]),
                          on_backoff=leaky_bucket_handler,
                          # No jitter as we want a constant value
                          jitter=None)
    @functools.wraps(fnc)
    def wrapper(*args, **kwargs):
        return fnc(*args, **kwargs)
    return wrapper

class Error(Exception):
    """Base exception for the API interaction module"""

class OutOfOrderIdsError(Error):
    """Raised if our expectation of ordering by ID is violated"""

class Stream():
    # Used for bookmarking and stream identification. Is overridden by
    # subclasses to change the bookmark key.
    name = None
    replication_method = 'INCREMENTAL'
    replication_key = 'updated_at'
    key_properties = ['id']
    # Controls which SDK object we use to call the API by default.
    replication_object = None
    # Status parameter override option
    status_key = None
    results_per_page = None

    def __init__(self):
        self.results_per_page = Context.get_results_per_page(RESULTS_PER_PAGE)

        # set request timeout
        self.request_timeout = get_request_timeout()

    def get_bookmark(self):
        bookmark = (singer.get_bookmark(Context.state,
                                        # name is overridden by some substreams
                                        self.name,
                                        self.replication_key)
                    or Context.config["start_date"])
        return utils.strptime_with_tz(bookmark)

    def get_since_id(self):
        return singer.get_bookmark(Context.state,
                                   # name is overridden by some substreams
                                   self.name,
                                   'since_id')

    def update_bookmark(self, bookmark_value, bookmark_key=None):
        # NOTE: Bookmarking can never be updated to not get the most
        # recent thing it saw the next time you run, because the querying
        # only allows greater than or equal semantics.
        singer.write_bookmark(
            Context.state,
            # name is overridden by some substreams
            self.name,
            bookmark_key or self.replication_key,
            bookmark_value
        )
        singer.write_state(Context.state)


    # This function can be overridden by subclasses for specialized API
    # interactions. If you override it you need to remember to decorate it
    # with shopify_error_handling to get 429 and 500 handling.
    @shopify_error_handling()
    def call_api(self, query_params):
        # set timeout
        self.replication_object.set_timeout(self.request_timeout)
        return self.replication_object.find(**query_params)

    def get_query_params(self, since_id, status_key, updated_at_min, updated_at_max):
        return {
            "since_id": since_id,
            "updated_at_min": updated_at_min,
            "updated_at_max": updated_at_max,
            "limit": self.results_per_page,
            status_key: "any"
        }

    def get_objects(self):
        updated_at_min = self.get_bookmark()

        stop_time = singer.utils.now().replace(microsecond=0)
        date_window_size = float(Context.config.get("date_window_size", DATE_WINDOW_SIZE))

        # Page through till the end of the resultset
        while updated_at_min < stop_time:
            # Bookmarking can also occur on the since_id
            since_id = self.get_since_id() or 1

            if since_id != 1:
                LOGGER.info("Resuming sync from since_id %d", since_id)

            # It's important that `updated_at_min` has microseconds
            # truncated. Why has been lost to the mists of time but we
            # think it has something to do with how the API treats
            # microseconds on its date windows. Maybe it's possible to
            # drop data due to rounding errors or something like that?
            updated_at_max = updated_at_min + datetime.timedelta(days=date_window_size)
            if updated_at_max > stop_time:
                updated_at_max = stop_time
            while True:
                status_key = self.status_key or "status"
                query_params = self.get_query_params(since_id,
                                                     status_key,
                                                     updated_at_min,
                                                     updated_at_max)

                with metrics.http_request_timer(self.name):
                    objects = self.call_api(query_params)

                for obj in objects:
                    if obj.id < since_id:
                        # This verifies the api behavior expectation we
                        # have that all results actually honor the
                        # since_id parameter.
                        raise OutOfOrderIdsError("obj.id < since_id: {} < {}".format(
                            obj.id, since_id))
                    yield obj

                # You know you're at the end when the current page has
                # less than the request size limits you set.
                if len(objects) < self.results_per_page:
                    # Save the updated_at_max as our bookmark as we've synced all rows up in our
                    # window and can move forward. Also remove the since_id because we want to
                    # restart at 1.
                    Context.state.get('bookmarks', {}).get(self.name, {}).pop('since_id', None)
                    self.update_bookmark(utils.strftime(updated_at_max))
                    break

                if objects[-1].id != max([o.id for o in objects]):
                    # This verifies the api behavior expectation we have
                    # that all pages are internally ordered by the
                    # `since_id`.
                    raise OutOfOrderIdsError("{} is not the max id in objects ({})".format(
                        objects[-1].id, max([o.id for o in objects])))
                since_id = objects[-1].id

                # Put since_id into the state.
                self.update_bookmark(since_id, bookmark_key='since_id')

            updated_at_min = updated_at_max

    def sync(self):
        """Yield's processed SDK object dicts to the caller.

        This is the default implementation. Get's all of self's objects
        and calls to_dict on them with no further processing.
        """
        for obj in self.get_objects():
            yield obj.to_dict()
